src/npg_porch/db/data_access.py

```diff
@@ -247,6 +250,49 @@ async def get_tasks(
         tasks = task_result.scalars().all()
         return [t.convert_to_model() for t in tasks]
 
+    async def get_ordered_tasks(self,
+                                filters: dict,
+                                limit: int,
+                                page: int,
+                                date_range: tuple[datetime.date, datetime.date] | None = None,
+                                ) -> list[TaskExpanded]:
+        """
+        Gets tasks in order of their most recent status change, with a provided
+        limit value, and an offset determined by which page is being displayed.
+
+        Can be filtered by pipeline and status. TODO: input and date.
+        """
+        latest_event = select(Event.task_id, samax(Event.time).label("changed"))\
```

Comment:
Would you consider printing (once) or logging the query this generates? I've gotten rusty on the sqlalchemy grammar.

Reply:
From logs:
```
INFO:     SELECT task.task_id, task.pipeline_id, task.job_descriptor, task.definition, task.state, task.prefix, task.created, pipeline_1.pipeline_id AS pipeline_id_1, pipeline_1.name, pipeline_1.repository_uri, pipeline_1.version
FROM task JOIN pipeline ON pipeline.pipeline_id = task.pipeline_id JOIN (SELECT event.task_id AS task_id, max(event.time) AS changed
FROM event GROUP BY event.task_id) AS anon_1 ON task.task_id = anon_1.task_id LEFT OUTER JOIN pipeline AS pipeline_1 ON pipeline_1.pipeline_id = task.pipeline_id ORDER BY anon_1.changed DESC
 LIMIT :param_1 OFFSET :param_2
```

Reply:
Yeah... this is giving the postgres optimizer nightmares.
Read the entire event table, read the entire task table. Both are high cardinality tables, and it gets progressively worse as Porch gets more and more used. Then sort it all so you can pick the top ones.

It's ok for trials you're doing now. For realistic high task throughput, it would probably be better to limit the number of discrete tasks by windowing the events to the last week or something, and add an index to the event table's time column. That or limit by the age since task creation (task table only), then bring in any event history from that.

Reply:
I don't know much about optimising db interaction, especially in postgres, but perhaps a `current` column in the event table would help?

Reply:
I have created a new [sub-task](https://jira.sanger.ac.uk/browse/NPG-2914) for this since it is not immediately necessary and will require discussion.